torch/_prims/__init__.py

```diff
@@ -0,0 +1,566 @@
+  they are all "nested" with each other. That is, they all
+  have the property that
+
+  stride[i] = stride[i+1] * shape[i+1]
+
+  for all i in [start, end - 1).
+  '''
+
+  _merge_dims_checks(a, start, end)
+  return _merge_dims_aten(a, start, end)
+
+def _split_dim_checks(a: Tensor, dim: int, outer_length: int):
+  assert isinstance(a, Tensor)
+  assert isinstance(dim, int)
+  assert isinstance(outer_length, int)
+
+  # Verifies dim is a valid idx
+  assert dim >=0 and dim < len(a.shape)
+
+  # Verifies outer_length is a valid length
+  assert outer_length >= 0
+
+  # Verifies the dim can be split with the specified lhs_length
+  inner_length = a.shape[dim] / outer_length
+  assert int(inner_length) == inner_length
+
+def _split_dim_aten(a: Tensor, dim: int, outer_length: int) -> Tensor:
+  inner_length = int(a.shape[dim] / outer_length)
+  new_shape = a.shape[0:dim] + (outer_length, inner_length) + a.shape[dim + 1:]
+
+  return a.view(new_shape)
+
+def split_dim(a: Tensor, dim: int, outer_length: int) -> Tensor:
+  '''
+  Creates a view of a with the given dimension (of length l) split
+  into two dimensions, with the outer of the two having
+  length outer_length and the inner of the two having computed
+  length inner_length such outer_length * inner_length = l.
+  '''
+
+  _split_dim_checks(a, dim, outer_length)
+  return _split_dim_aten(a, dim, outer_length)
+
+# Note: allows dimensions to be specified redundantly
+def _squeeze_checks(a: Tensor, dimensions: Sequence):
+  assert isinstance(a, Tensor)
+
+  for idx in dimensions:
+    _validate_idx(a.shape, idx)
+    assert a.shape[idx] == 1
+
+def _squeeze_aten(a: Tensor, dimensions: Sequence) -> Tensor:
+  for idx in dimensions:
+    a = torch.squeeze(a, dim=idx)
+
+  return a
+
+def squeeze(a: Tensor, dimensions: Sequence) -> Tensor:
+  '''
+  Creates a view of a with the specified dimensions of
+  length one removed.
+  '''
+
+  _squeeze_checks(a, dimensions)
+  return _squeeze_aten(a, dimensions)
+
+#
+# Shape operations
+#
+def collapse(a: Tensor, start: int, end: int) -> Tensor:
+  '''
+  Wrapper around reshape that collapses a span of dimensions.
+
+  See merge_dims for the corresponding view operation.
+  '''
+
+  dim_length = 1
+  for idx in range(start, end):
+    dim_length = dim_length * a.shape[idx]
+
+  new_shape = a.shape[0:start] + (dim_length,) + a.shape[end:]
+  return reshape(a, new_shape)
+
+def _concatenate_checks(tensors: Sequence[Tensor], dim: int):
+  assert len(tensors) > 0
+  assert dim >= 0
+
+  shape = tensors[0].shape
+  assert dim < len(shape)
+
+  _check_same_dtype(*tensors)
+
+  # Verifies same shape (except in the concat dimension)
+  for tensor in tensors:
+    for idx, lengths in enumerate(zip(shape, tensor.shape)):
+      common_length, length = lengths
+      if idx == dim:
+        continue
+      assert length == common_length
+
+def _concatentate_aten(tensors: Sequence[Tensor], dim: int) -> Tensor:
+  return torch.cat(tensors, dim)
```

Comment:
Do we want to be so programatic as to define aliases? Is this uniform naming convention used anywhere?